KIPs/kip-draft_kgp6.md

```diff
@@ -0,0 +1,652 @@
+                );
+            } else {
+                revert("min required admins should approve");
+            }
+        }
+    }
+
+    /**
+     * @dev sets the status of the contract to Finalize. Once finalized the storage data
+     * of the contract cannot be modified
+     * @param _memo is the result of the rebalance after executing successfully in the core.
+     */
+    function finalizeContract(
+        string memory _memo
+    ) public onlyOwner onlyAtStatus(Status.Approved) {
+        memo = _memo;
+        status = Status.Finalized;
+        emit Finalized(memo, status);
+    }
+
+    /**
+     * @dev resets all storage values to empty objects except targetBlockNumber
+     */
+    function reset() public onlyOwner {
+        //reset cannot be called at Finalized status or after target block.number
+        require(
+            ((status != Status.Finalized) &&
+                (block.number < rebalanceBlockNumber)),
+            "Contract is finalized, cannot reset values"
+        );
+
+        //`delete` keyword is used to set a storage variable or a dynamic array to its default value.
+        delete retirees;
+        delete newbies;
+        delete memo;
+        status = Status.Initialized;
+    }
+
+    //Getters
+    /**
+     * @dev to get retired details by retiredAddress
+     * @param _retiredAddress is the address of the retired
+     */
+    function getRetired(
+        address _retiredAddress
+    ) public view returns (address, address[] memory) {
+        require(retiredExists(_retiredAddress), "Retired does not exist");
+        uint256 index = getRetiredIndex(_retiredAddress);
+        Retired memory retired = retirees[index];
+        return (retired.retired, retired.approvers);
+    }
+
+    /**
+     * @dev check whether retiredAddress is registered
+     * @param _retiredAddress is the address of the retired
+     */
+    function retiredExists(address _retiredAddress) public view returns (bool) {
+        require(_retiredAddress != address(0), "Invalid address");
+        for (uint8 i = 0; i < retirees.length; i++) {
+            if (retirees[i].retired == _retiredAddress) {
+                return true;
+            }
+        }
+    }
+
+    /**
+     * @dev get index of the retired in the retirees array
+     * @param _retiredAddress is the address of the retired
+     */
+    function getRetiredIndex(
+        address _retiredAddress
+    ) public view returns (uint256) {
+        for (uint256 i = 0; i < retirees.length; i++) {
+            if (retirees[i].retired == _retiredAddress) {
+                return i;
+            }
+        }
+        revert("Retired does not exist");
+    }
+
+    /**
+     * @dev to calculate the sum of retirees balances
+     * @return retireesBalance the sum of balances of retireds
+     */
+    function sumOfRetiredBalance() public view returns (uint256 retireesBalance) {
+        for (uint8 i = 0; i < retirees.length; i++) {
```

Comment:
This might be not a big deal, but if you want to use uint8, you need to limit the length of this array less than uint8_max.

Reply:
Also there is some inconsistencies of using index of this array.

https://github.com/klaytn/kips/pull/104/files#diff-ab7a71791e7f914861f0980b3c2d95e2d5794bf10bc2548647a8fcc6d25cddbfR272

https://github.com/klaytn/kips/pull/104/files#diff-ab7a71791e7f914861f0980b3c2d95e2d5794bf10bc2548647a8fcc6d25cddbfR526-R528

Reply:
fixed uint inconsistencies